test/cli/harness/cluster_node.go

```diff
@@ -0,0 +1,391 @@
+package harness
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/ipfs/kubo/config"
+	serial "github.com/ipfs/kubo/config/serialize"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/multiformats/go-multiaddr"
+)
+
+var log = logging.Logger("testharness")
+
+type Node struct {
+	ID  int
+	Dir string
+
+	APIListenAddr multiaddr.Multiaddr
+	SwarmAddr     multiaddr.Multiaddr
+	EnableMDNS    bool
+
+	IPFSBin string
+	Runner  *Runner
+
+	daemon *RunResult
+}
+
+func BuildNode(ipfsBin, baseDir string, id int) Node {
+	dir := filepath.Join(baseDir, strconv.Itoa(id))
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		panic(err)
+	}
+
+	env := environToMap(os.Environ())
+	env["IPFS_PATH"] = dir
+
+	return Node{
+		ID:      id,
+		Dir:     dir,
+		IPFSBin: ipfsBin,
+		Runner: &Runner{
+			Env: env,
+			Dir: dir,
+		},
+	}
+}
+
+func (n *Node) ReadConfig() *config.Config {
+	cfg, err := serial.Load(filepath.Join(n.Dir, "config"))
+	if err != nil {
+		panic(err)
+	}
+	return cfg
+}
+
+func (n *Node) WriteConfig(c *config.Config) {
+	err := serial.WriteConfigFile(filepath.Join(n.Dir, "config"), c)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (n *Node) UpdateConfig(f func(c *config.Config)) {
+	cfg := n.ReadConfig()
+	f(cfg)
+	n.WriteConfig(cfg)
+}
+
+func (n *Node) IPFS(args ...string) RunResult {
+	res := n.RunIPFS(args...)
+	n.Runner.AssertNoError(res)
+	return res
+}
+
+func (n *Node) PipeStrToIPFS(s string, args ...string) RunResult {
+	return n.PipeToIPFS(strings.NewReader(s), args...)
+}
+
+func (n *Node) PipeToIPFS(reader io.Reader, args ...string) RunResult {
+	res := n.RunPipeToIPFS(reader, args...)
+	n.Runner.AssertNoError(res)
+	return res
+}
+
+func (n *Node) RunPipeToIPFS(reader io.Reader, args ...string) RunResult {
+	return n.Runner.Run(RunRequest{
+		Path:    n.IPFSBin,
+		Args:    args,
+		CmdOpts: []CmdOpt{RunWithStdin(reader)},
+	})
+}
+
+func (n *Node) RunIPFS(args ...string) RunResult {
+	return n.Runner.Run(RunRequest{
+		Path: n.IPFSBin,
+		Args: args,
+	})
+}
+
+// Init initializes and configures the IPFS node, after which it is ready to run.
+func (n *Node) Init(ipfsArgs ...string) {
+	// 	// h.MustRunIPFS("init", "--profile=test")
+	// 	// h.Mkdirs("mountdir", "ipfs", "ipns")
+	// 	// h.SetIPFSConfig("Mounts.IPFS", h.IPFSMountpoint)
+	// 	// h.SetIPFSConfig("Mounts.IPNS", h.IPNSMountpoint)
+
+	// 	// configPath := filepath.Join(h.IPFSPath, "config")
+	// 	// b, err := ioutil.ReadFile(configPath)
+	// 	// if err != nil {
+	// 	// 	log.Fatalf("reading config file from %s: %s", configPath, err)
+	// 	// }
+	// 	// if len(b) == 0 {
+	// 	// 	log.Panicf("expected non-empty config at %s", configPath)
+	// 	// }
```

Comment:
remove

Reply:
Also I'm not sure what the test profile does, should we use that here?

Reply:
test profile:

```
			c.Addresses.API = Strings{"/ip4/127.0.0.1/tcp/0"}
			c.Addresses.Gateway = Strings{"/ip4/127.0.0.1/tcp/0"}
			c.Addresses.Swarm = []string{
				"/ip4/127.0.0.1/tcp/0",
			}

			c.Swarm.DisableNatPortMap = true

			c.Bootstrap = []string{}
			c.Discovery.MDNS.Enabled = false
			return nil

```

we don't really need this, the test harness already sets up this config, and that's definitely more clear then having to look this up somewhere else